reasoner/resolution/ResolverRegistry.java

```diff
@@ -18,51 +18,78 @@
 
 package grakn.core.reasoner.resolution;
 
+import grakn.common.collection.Pair;
 import grakn.common.concurrent.actor.Actor;
 import grakn.common.concurrent.actor.EventLoopGroup;
-import grakn.core.reasoner.resolution.framework.Answer;
+import grakn.core.logic.Rule;
+import grakn.core.logic.concludable.ConjunctionConcludable;
+import grakn.core.pattern.Conjunction;
+import grakn.core.pattern.equivalence.AlphaEquivalence;
+import grakn.core.reasoner.resolution.answer.MappingAggregator;
+import grakn.core.reasoner.resolution.framework.ResolutionAnswer;
 import grakn.core.reasoner.resolution.resolver.ConcludableResolver;
 import grakn.core.reasoner.resolution.resolver.RootResolver;
 import grakn.core.reasoner.resolution.resolver.RuleResolver;
+import graql.lang.pattern.variable.Reference;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.HashMap;
-import java.util.List;
+import java.util.Map;
+import java.util.Optional;
 import java.util.function.Consumer;
-
 public class ResolverRegistry {
 
     Logger LOG = LoggerFactory.getLogger(ResolverRegistry.class);
-    private final HashMap<Long, Actor<ConcludableResolver>> concludables;
-    private final HashMap<List<Long>, Actor<RuleResolver>> rules;
+    private final HashMap<ConjunctionConcludable<?, ?>, Actor<ConcludableResolver>> concludableActorsMap;
+    private final HashMap<Rule, Actor<RuleResolver>> rules;
     private final Actor<ResolutionRecorder> resolutionRecorder;
-    private final EventLoopGroup elg;
+    private EventLoopGroup elg;
 
     public ResolverRegistry(EventLoopGroup elg) {
         this.elg = elg;
-        concludables = new HashMap<>();
+        concludableActorsMap = new HashMap<>();
         rules = new HashMap<>();
         resolutionRecorder = Actor.create(elg, ResolutionRecorder::new);
     }
 
-    public Actor<ConcludableResolver> registerConcludable(Long pattern, List<List<Long>> rules, long traversalSize) {
-        LOG.debug("Register retrieval for concludable actor: '{}'", pattern);
-        return concludables.computeIfAbsent(pattern, (p) -> Actor.create(elg, self -> new ConcludableResolver(self, p, rules, traversalSize)));
+    public Pair<Actor<ConcludableResolver>, Map<Reference.Name, Reference.Name>> registerConcludable(ConjunctionConcludable<?, ?> concludable) {
+        LOG.debug("Register retrieval for concludable actor: '{}'", concludable.conjunction());
+
+        final Optional<Pair<ConjunctionConcludable<?, ?>, AlphaEquivalence>> alphaEquivalencePair = concludableActorsMap.keySet().stream()
+                .map(k -> new Pair<ConjunctionConcludable<?, ?>, AlphaEquivalence>(k, k.alphaEquals(concludable)))
+                .filter(p -> p.second().isValid()).findAny();
```

Comment:
Here we're interested in registering deduplicated Concludables, so we want alpha equivalence. We iterate over all concludables already registered. If any is alpha equivalent then we add it and it's AlphaEquivalence to a pair, such that both can be accessed in the next step. Perhaps we can find a clearer flow for this, perhaps just a for loop with a `break` on finding a match?

Reply:
I've refactored this to use a for loop now - much cleaner